src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElement.java

```diff
@@ -111,23 +111,23 @@ public XpathFilterElement(Pattern files, Pattern checks, Pattern message,
         }
         else {
             filePattern = files.pattern();
-            fileRegexp = files;
+            fileRegexp = Pattern.compile(filePattern);
```

Comment:
This change makes no sense to de-pattern a pattern then to re-pattern it right back again.

Reply:
@Kevin222004 , try to remove fields of type String that are string representation of compiled pattern.
all their usage in hashCode/equals change to `xxxx.pattern()`.

Reply:
@Kevin222004 , can you split this PR in two ? to let us merge what is easy and discuss/review deeply only narrow set of cases

Reply:
for null cases make method `getPatternSafely(pattern) { if (pattern == null) { return null} else {return patter.patern()}}
﻿` 
to make `xxxx.pattern()` without exceptions.

Reply:
Done